src/Functions/CountSubstringsImpl.h

```diff
@@ -62,7 +62,7 @@ struct CountSubstringsImpl
         while (pos < end && end != (pos = searcher.search(pos, end - pos)))
         {
             /// Determine which index it refers to.
-            while (begin + haystack_offsets[i] <= pos)
+            while (i + 1 < input_rows_count && begin + haystack_offsets[i] <= pos)
```

Comment:
Can we do something like this or `input_rows_count` can be 0 ?
```suggestion
            while (i < input_rows_count - 1 && begin + haystack_offsets[i] <= pos)
```

`input_rows_count` is basically `const`, and so `input_rows_count` - 1